### ISSUE TITLE
Get rid of TempNupkgFile

### ISSUE THREAD

**Issue by sbespalov** [strongbox/strongbox#1146]
1. Move `TempNupkgFile.loadNuspec` method logic into separate class `org.carlspring.strongbox.nuget.NuspecExtractor`.
2. Use `RepositoryPathResolver` to resolve the `RepositoryPath` and then store it with `ArtifactManagementService.validateAndStore` method directly from the `InputStream` we have in request.
3. Get the checksums from `ArtifactEntry` (you can get the `ArtifactEntry` wtih `RepositoryPath.getArtifactEntry()` method).
4. Store every checksum with `ArtifactManagementService.validateAndStore` method.
5. Extract **nuspec** file with `NuspecExtractor` and store it with `ArtifactManagementService.validateAndStore`.
6. Store **nuspec** checksum same way as №3.

Note:
This issue requires the #1005 to be done first.

**Comment by carlspring** [comment]
Hi @ogryniuk ,

Thanks for offering to help! Do you understand what needs to be done? Do you need further clarifications? Are you familiar with NuGet?

**Comment by ogryniuk** [comment]
@carlspring 

I think I understand what needs to be done. 
For now I do not need further information. The instructions are quite detailed but if I struggle, I will ask questions. 
No, I am not familiar with NuGet. However, I just looked up on the Net what it serves for and, based on the description of the issue itself, the task does not require much knowledge of NuGet per se

**Comment by carlspring** [comment]
Okay!

If you  need help, @sbespalov and @kalski would know more.

**Comment by kalski** [comment]
@carlspring I can take this one, please assign it to me.

**Comment by kalski** [comment]
@sbespalov , regarding these:

> 1. Move `TempNupkgFile.loadNuspec` method logic into separate class `org.carlspring.strongbox.nuget.NuspecExtractor`.
> 
> 2. Use `RepositoryPathResolver` to resolve the `RepositoryPath` and then store it with `ArtifactManagementService.validateAndStore` method directly from the `InputStream` we have in request.

To read the Nuspec necessary to build the file name to be used when resolving the `RepositoryPath` in 2. I have to run `NuspecExtractor#loadNuspec` from 1. This method receives an `InputStream` - the request `InputStream`. Once the `InputStream` is read for this, 2. can no longer use it.

So I see 2 options here for the request `InputStream`:
1. Write it to a temporary file - basically that's what `TempNupkgFile` does. Well it also calculates the checksum while writing this file.
2. Decorate it with logic to read the Nuspec. This way when reading the `InputStream` to save it in a file "near the repository" we'll also have the Nuspec extracted. Once Nuspec is read we can resolve the `RepositoryPath` and move the file there.

I think 1. is simpler and not that slower so I prefer it.
2. suggests exposing "nearArtifactLocation" from repositories. Otherwise it just doesn't make sense.

As for writing the checksum(s) - I noticed [ArtifactManagementService#writeArtifact](https://github.com/strongbox/strongbox/blob/5ec54892d37c0b3f610123f185bed108b31f5ced/strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/ArtifactManagementService.java#L177) actually does this when calling `ArtifactManagementService.validateAndStore` for the Nupkg itself. Since we have no checksum in the request I don't think it makes sense to use the one we generate while writing the Nupkg to validate itself.

As a conclusion I would say I don't think `TempNupkgFile` is that bad of an abstraction - it creates a temp file to expose its `InputStream` (we pass the request one) and also generates its checksum which we already use in `ArtifactManagementService#validateAndStore` to validate against the checksum computed while writing the Nupkg to the repository. We also already use `RepositoryPathResolver#resolve` to resolve the path for the Nupkg artifact.

**Comment by sbespalov** [comment]
@kalski seems you right about file name issue, so we can to do it a bit more complex way. 

Regarding your proposals the thing is that we already write `InputStream` into temporary location using `InputStream` decorators from `StorageFileSystemProvider`, we also able to have Layout specific decorators within concrete `LayoutFileSystemProvider` implementations, so I think it will be good to have such decorator for Nuget to expose Nuspec from underlying `InputStream` and then write the package where needed.
In this case it will also need to have lazy `ReoisitoryPath` decorator, which will require all methods/operations to be available only after temporary location write complete.

All this will help to avoid extra read/write of incoming package file, so it will be same fast as regular artifact write.

>As for writing the checksum(s) - I noticed ArtifactManagementService#writeArtifact actually does this when calling ArtifactManagementService.validateAndStore for the Nupkg itself. Since we have no checksum in the request I don't think it makes sense to use the one we generate while writing the Nupkg to validate itself.

yeah, we probably need to drop this part as it already implemented within `ArtifactManagementService` and works like required, we only need a test case to prevent Nuget functionality regress.

**Comment by carlspring** [comment]
@sbespalov ,

Is this task still something we would like to look into?

**Comment by sbespalov** [comment]
yes, but we probably need to update the description
will do it

**Comment by carlspring** [comment]
@sbespalov ,

What would you like to update in the description? Would you mind looking into it?

**Comment by sbespalov** [comment]
@carlspring probably we should put this "on hold" to better times.
